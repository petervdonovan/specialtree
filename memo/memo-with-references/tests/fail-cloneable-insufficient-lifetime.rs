use memo_with_references::memo;

struct DataWithLifetime<'b> {
    data: &'b str,
}

// This should fail because DataWithLifetime<'b> cannot outlive 'a
// when 'b is not bound to outlive 'a
#[memo('a)]
fn test_insufficient_lifetime<'a, 'b>(x: DataWithLifetime<'b>) -> &'a DataWithLifetime<'b> {
    // Even after cloning DataWithLifetime<'b>, the clone still has the 'b lifetime
    // which may not be long enough for the cache's 'a lifetime
    unsafe { std::mem::transmute(&x) }
}

fn main() {}
