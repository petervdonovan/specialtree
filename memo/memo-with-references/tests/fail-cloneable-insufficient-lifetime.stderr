error: lifetime may not live long enough
  --> tests/fail-cloneable-insufficient-lifetime.rs:9:1
   |
9  | #[memo('a)]
   | ^^^^^^^^^^^ requires that `'b` must outlive `'static`
10 | fn test_insufficient_lifetime<'a, 'b>(x: DataWithLifetime<'b>) -> &'a DataWithLifetime<'b> {
   |                                   -- lifetime `'b` defined here
   |
   = note: this error originates in the attribute macro `memo` (in Nightly builds, run with -Z macro-backtrace for more info)
